### PR TITLE
feat: customizable model categories

### PR DESCRIPTION
**Pull request by gitlawr**
https://github.com/gpustack/gpustack/issues/810
https://github.com/gpustack/gpustack/issues/598

Deprecate previous category fields and add a unified categories field. 
It defaults to Auto in the UI and users can custom in advanced settings.

### CODE REVIEW

File: gpustack/migrations/versions/2024_12_26_1409-e6bf9e067296_add_model_categories.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    with op.batch_alter_table('models', schema=None) as batch_op:
+        batch_op.add_column(sa.Column('categories', sa.JSON(), nullable=True))
+    
+    with op.batch_alter_table('models', schema=None) as batch_op:
+        connection = batch_op.get_bind()
+        categories_case = sa.case(
+                (sa.column('reranker') == True, sa.literal(json.dumps(['reranker']))),
+                (sa.column('embedding_only') == True, sa.literal(json.dumps(['embedding']))),
+                (sa.column('image_only') == True, sa.literal(json.dumps(['image']))),
+                (sa.column('speech_to_text') == True, sa.literal(json.dumps(['speech_to_speech']))),
```
Comment (aiwantaozi) [inline gpustack/migrations/versions/2024_12_26_1409-e6bf9e067296_add_model_categories.py]: speech_to_speech -> speech_to_text

File: gpustack/migrations/versions/2024_12_26_1409-e6bf9e067296_add_model_categories.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+# revision identifiers, used by Alembic.
+revision: str = 'e6bf9e067296'
+down_revision: Union[str, None] = '004c73a5c09e'
+branch_labels: Union[str, Sequence[str], None] = None
+depends_on: Union[str, Sequence[str], None] = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (aiwantaozi) [inline gpustack/migrations/versions/2024_12_26_1409-e6bf9e067296_add_model_categories.py]: The comment should be removed after we adjusted.

File: gpustack/migrations/versions/2024_12_26_1409-e6bf9e067296_add_model_categories.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                (sa.column('reranker') == True, sa.func.to_json(['reranker'])),
+                (sa.column('embedding_only') == True, sa.func.to_json(['embedding'])),
+                (sa.column('image_only') == True, sa.func.to_json(['image'])),
+                (sa.column('speech_to_text') == True, sa.func.to_json(['speech_to_speech'])),
+                (sa.column('text_to_speech') == True, sa.func.to_json(['text_to_speech'])),
+                else_=sa.literal([])
+            )
+        
+        models_table = sa.table('models', sa.column('categories', sa.JSON))
+        # sa.table('models')
```
Comment (aiwantaozi) [inline gpustack/migrations/versions/2024_12_26_1409-e6bf9e067296_add_model_categories.py]: What is this comment for?

**Comment by gitlawr** [comment]
Updated

**Review by aiwantaozi — APPROVED** [review]
LGTM

**Pull request merged**